### PR TITLE
Dipy newversion

### PR DESCRIPTION
**Pull request by loftusa**
don't merge yet, just PRing to see the diff

### CODE REVIEW

**Review by dPys — DISMISSED** [review]

**Review by Lawreros — DISMISSED** [review]
Really like how clear the outputs are now

**Comment by loftusa** [comment]
Alright, given that this check is successful, I'm going to merge into staging, even though MNI space still isn't working, because

a) we don't use MNI space much
b) therefore, it might be more trouble than it's worth to get it fixed for now
c) there are other changes in this branch that are useful for us to have (e.g., improved logs, dipy 1.0 on native space, tracebacking immediately rather than 10 steps later if necessary files don't exist with the new decorator)

before merging into deploy, we need to either fix MNI space or deprecate it.

**Review by dPys — CHANGES_REQUESTED** [review]
Please incorporate the following changes to DSN:

```
    # SyN FA->Template
    [mapping, affine_map, warped_fa] = regutils.wm_syn(template_path, fa_path, dsn_dir)

    tractogram = load_tractogram(streams, fa_img, to_space=Space.RASMM, shifted_origin=True, bbox_valid_check=False)
    streamlines = tractogram.streamlines
    warped_fa_img = nib.load(warped_fa)

    streams_in_curr_grid = transform_streamlines(streamlines, warped_fa_img.affine)

    ref_grid_aff = vox_size*np.eye(4)
    ref_grid_aff[3][3] = 1

    # Deform streamlines
    displacements = values_from_volume(mapping.get_forward_field(), streams_in_curr_grid, ref_grid_aff)

    deformed_streams = [sum(d, s) for d, s in zip(displacements, streams_in_curr_grid)]

    # Create isocenter mapping where we anchor the origin transformation affine
    # to the corner of the FOV by scaling x, y, z offsets according to a multiplicative
    # van der Corput sequence with a base value equal to the voxel resolution
    def vdc(n, base=vox_size):
        vdc, denom = 0, 1
        while n:
            denom *= base
            n, remainder = divmod(n, base)
            vdc += remainder / denom
        return vdc

    [x_mul, y_mul, z_mul] = [vdc(i) for i in range(1, 4)]

    adjusted_affine = affine_map.affine.copy()
    adjusted_affine[0][3] = -adjusted_affine[0][3]*x_mul
    adjusted_affine[1][3] = -adjusted_affine[1][3]*y_mul
    adjusted_affine[2][3] = -adjusted_affine[2][3]*z_mul

    streams_isocentered = transform_streamlines(deformed_streams, np.linalg.inv(adjusted_affine))

    streams_final = transform_streamlines(streams_isocentered, np.linalg.inv(warped_fa_img.affine))

    # Remove streamlines outside brain
    streams_final_filt = Streamlines(utils.target_line_based(streams_final, np.eye(4), brain_mask, include=True))

    # Remove streamlines with negative voxel indices
    lin_T, offset = _mapping_to_voxel(np.eye(4))
    streams_final_filt_final = []
    for sl in streams_final_filt:
        inds = np.dot(sl, lin_T)
        inds += offset
        if not inds.min().round(decimals=6) < 0:
            streams_final_filt_final.append(sl)

    # Save streamlines
    stf = StatefulTractogram(streams_final_filt_final, reference=warped_fa_img, space=Space.RASMM, shifted_origin=True)
    stf.remove_invalid_streamlines()
    streams_final_filt_final = stf.streamlines
    save_tractogram(stf, streams_mni, bbox_valid_check=True)
```

**Comment by loftusa** [comment]
> Please incorporate the following changes to DSN:
> 
> ```
>     # SyN FA->Template
>     [mapping, affine_map, warped_fa] = regutils.wm_syn(template_path, fa_path, dsn_dir)
> 
>     tractogram = load_tractogram(streams, fa_img, to_space=Space.RASMM, shifted_origin=True, bbox_valid_check=False)
>     streamlines = tractogram.streamlines
>     warped_fa_img = nib.load(warped_fa)
> 
>     streams_in_curr_grid = transform_streamlines(streamlines, warped_fa_img.affine)
> 
>     ref_grid_aff = vox_size*np.eye(4)
>     ref_grid_aff[3][3] = 1
> 
>     # Deform streamlines
>     displacements = values_from_volume(mapping.get_forward_field(), streams_in_curr_grid, ref_grid_aff)
> 
>     deformed_streams = [sum(d, s) for d, s in zip(displacements, streams_in_curr_grid)]
> 
>     # Create isocenter mapping where we anchor the origin transformation affine
>     # to the corner of the FOV by scaling x, y, z offsets according to a multiplicative
>     # van der Corput sequence with a base value equal to the voxel resolution
>     def vdc(n, base=vox_size):
>         vdc, denom = 0, 1
>         while n:
>             denom *= base
>             n, remainder = divmod(n, base)
>             vdc += remainder / denom
>         return vdc
> 
>     [x_mul, y_mul, z_mul] = [vdc(i) for i in range(1, 4)]
> 
>     adjusted_affine = affine_map.affine.copy()
>     adjusted_affine[0][3] = -adjusted_affine[0][3]*x_mul
>     adjusted_affine[1][3] = -adjusted_affine[1][3]*y_mul
>     adjusted_affine[2][3] = -adjusted_affine[2][3]*z_mul
> 
>     streams_isocentered = transform_streamlines(deformed_streams, np.linalg.inv(adjusted_affine))
> 
>     streams_final = transform_streamlines(streams_isocentered, np.linalg.inv(warped_fa_img.affine))
> 
>     # Remove streamlines outside brain
>     streams_final_filt = Streamlines(utils.target_line_based(streams_final, np.eye(4), brain_mask, include=True))
> 
>     # Remove streamlines with negative voxel indices
>     lin_T, offset = _mapping_to_voxel(np.eye(4))
>     streams_final_filt_final = []
>     for sl in streams_final_filt:
>         inds = np.dot(sl, lin_T)
>         inds += offset
>         if not inds.min().round(decimals=6) < 0:
>             streams_final_filt_final.append(sl)
> 
>     # Save streamlines
>     stf = StatefulTractogram(streams_final_filt_final, reference=warped_fa_img, space=Space.RASMM, shifted_origin=True)
>     stf.remove_invalid_streamlines()
>     streams_final_filt_final = stf.streamlines
>     save_tractogram(stf, streams_mni, bbox_valid_check=True)
> ```

@dPys 

Where? Just replace `gen_reg.direct_streamline_norm`?

The code you gave doesn't return `streamlines_mni`, which is used in tractography for `ndmg_dwi_pipeline`.

What's the context?
Might be easier if you just make a PR

**Comment by dPys** [comment]
Yep, probs. I'll address this in next week's ndmg meeting when I'm back in town!
@dPys

**Comment by loftusa** [comment]
@dPys ok, re-requesting review to merge then - we can add these DSN changes in a different branch

File: ndmg/graph/gen_graph.py [hunk 1]
```diff
@@ -292,20 +292,9 @@ def save_graph(self, graphname, fmt="igraph"):
...
-                    graphname,
-                    self.g,
-                    comments="",
-                    delimiter=",",
-                    header=",".join([str(n) for n in self.n_ids]),
-                )
-            else:
-                raise ValueError("Unsupported Modality.")
+            nx.write_weighted_edgelist(
+                self.g, graphname, delimiter=" ", encoding="utf-8"
```
Comment (Lawreros) [inline ndmg/graph/gen_graph.py:296]: don't we want to make the edgelists comma delimited? (Based on an Issue you assigned to me that I haven't done yet)

**Review by Lawreros — APPROVED** [review]

File: ndmg/graph/gen_graph.py (hunk 1 above)
Comment (loftusa) [reply ndmg/graph/gen_graph.py:296]: Yep! That'll likely be in a future branch - we're going to basically rewrite this function anyway, cause we don't care about having any other modalities.

Could change it now, but once we do that I'll have to rewrite some of `graphutils` to account for it, and make sure nothing else gets broken, so I want to do that formally in an independent branch

**Pull request merged**